contracts/evm/contracts/Oracle.sol

```diff
@@ -0,0 +1,258 @@
+
+    mapping(bytes32 => bytes32[]) public oracleSourcesStorage;
+    mapping(bytes32 => mapping(bytes32 => SpotEntry)) public spotEntryStorage;
+    mapping(bytes32 => Checkpoint[]) public checkpoints;
+    mapping(bytes32 => uint256) public checkpointIndex;
+    uint256 sourcesThreshold = 1;
+
+    uint256 constant BACKWARD_TIMESTAMP_BUFFER = 3600;
+    uint256 constant FORWARD_TIMESTAMP_BUFFER = 900;
+
+    constructor(
+        address _publisherRegistry,
+        Currency[] memory _currencies,
+        Pair[] memory _pairs
+    ) {
+        publisherRegistry = IPublisherRegistry(_publisherRegistry);
+        for (uint256 i = 0; i < _currencies.length; i++) {
+            currencies[_currencies[i].id] = _currencies[i];
+        }
+        for (uint256 i = 0; i < _pairs.length; i++) {
+            pairs[_pairs[i].id] = _pairs[i];
+        }
+    }
+
+    function setSourcesThreshold(uint256 threshold) external onlyOwner {
+        sourcesThreshold = threshold;
+    }
+
+    function updatePublisherRegistryAddress(
+        IPublisherRegistry newPublisherRegistryAddress
+    ) external onlyOwner {
+        publisherRegistry = newPublisherRegistryAddress;
+    }
+
+    function publishSpotEntry(SpotEntry calldata spotEntry) public {
+        _validateSenderForSource(spotEntry.base, msg.sender);
+        SpotEntry memory _latest = spotEntryStorage[spotEntry.pairId][
+            spotEntry.base.source
+        ];
+        _validateTimestamp(_latest, spotEntry);
+        spotEntryStorage[spotEntry.pairId][spotEntry.base.source] = spotEntry;
+
+        emit SubmittedSpotEntry(spotEntry);
+    }
+
+    function setCheckpoint(bytes32 pairId, AggregationMode aggregationMode)
+        public
+    {
+        bytes32[] memory sources = oracleSourcesStorage[pairId];
+        (
+            uint256 value,
+            ,
+            uint256 lastUpdatedTimestamp,
+            uint256 numSourcesAggregated
+        ) = getSpot(pairId, aggregationMode, sources);
+
+        require(
+            sourcesThreshold <= numSourcesAggregated,
+            "Does not meet threshold for aggreagated sources"
+        );
+
+        if (checkpointIndex[pairId] > 0) {
+            Checkpoint memory currentCheckpoint = checkpoints[pairId][
+                checkpointIndex[pairId] - 1
+            ];
+            require(
+                currentCheckpoint.timestamp < lastUpdatedTimestamp,
+                "stale"
+            );
+        }
+        Checkpoint memory newCheckpoint = Checkpoint(
+            lastUpdatedTimestamp,
+            value,
+            aggregationMode,
+            numSourcesAggregated
+        );
+
+        checkpointIndex[pairId]++;
+        checkpoints[pairId].push(newCheckpoint);
+
+        return ();
+    }
+
+    function publishSpotEntries(SpotEntry[] calldata spotEntries) public {
+        for (uint256 i = 0; i < spotEntries.length; i++) {
+            publishSpotEntry(spotEntries[i]);
+        }
+    }
```

Comment:
After months of recursion for loops have never looked as beautiful

Reply:
yeah it's so much less painful